### ISSUE TITLE
tfp.distributions.JointDistributionSequential method sample() is not working as expected

### ISSUE THREAD

**Issue by xiaolong1979** [tensorflow/probability#1347]
Hi, I have similar issue before, but with sample() method. While sample() is working right but sample(2) (for multiple samples) does not work. Could you help to check? Thanks. See codes below:

import pandas as pd
import numpy as np
from pprint import pprint
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

import tensorflow as tf
import tensorflow_probability as tfp
from tensorflow_probability import distributions as tfd
tfd = tfp.distributions
tfb = tfp.bijectors

### create a class with fixed and random effect
class BayesianWLS(tf.Module):
    def __init__(self, num_targets,num_subjects,num_sources):
        ### setup random effect parameters
        self.mu_nu = tf.Variable(0, dtype=tf.float32,name="mean_nu") # mean of random subject effect nu
        self.std_nu = tfp.util.TransformedVariable(0.01, bijector=tfb.Exp(), name="std_nu")  # std in random subject effect nu
        self.std_eta = tfp.util.TransformedVariable(0.01, bijector=tfb.Exp(), name="std_eta")  # std in random subject slope effect eta
        self.std_delta = tfp.util.TransformedVariable(0.01, bijector=tfb.Exp(), name="std_delta")  # std in random strata slope effect eta
        # Set up fixed effects and other parameter
        self.beta =tf.Variable(tf.zeros(shape=(1,num_targets),dtype=tf.float32), shape=(1,num_targets), name="beta") #  beta in eq
        c_ones=int(num_targets*(num_targets+1)/2) ### for triangle matrix with 1
        self.bcumsum=tf.dtypes.cast(tf.transpose(tfp.math.fill_triangular([1]*c_ones)), tf.float32)
        ### set up error std
        self.std_f=tfp.util.TransformedVariable( 0.01+tf.zeros([1+num_sources, 1], tf.float32), 
                                                bijector=tfb.Exp())
 
        
    def __call__(self, X,source,treatment_group):
        ### get mixed effect means
        Bcov=tf.concat([tf.cast(treatment_group, tf.float32), X],axis=1)
        alpha_X=tf.math.exp(tf.keras.layers.Dense(1)(Bcov)) ### exp(alpha*X)
        ### get treatment and source specific error stds
        tx_source=tf.concat([tf.cast(treatment_group, tf.float32), tf.cast(source, tf.float32)],axis=1)
        scale_diag=tf.repeat(tf.exp(tf.matmul(tx_source, self.std_f)),num_targets,axis=1)
        model=tfd.JointDistributionSequential([
        # random effect
        tfd.Sample(
            tfd.Normal(loc=self.mu_nu, scale=self.std_nu),
            sample_shape=(num_subjects, 1)),  #nu=
        tfd.Sample(
            tfd.Gamma(concentration=1/(self.std_eta)**2, rate=1/(self.std_eta)**2),
            sample_shape=(num_subjects,1)),  ## eta
        tfd.Sample(tfd.Uniform(low=0.1, high=1.3),
                    sample_shape=(1,num_sources)),  ### std_delta
        lambda std_delta: tfd.Independent(
            tfd.Gamma(concentration=1 / std_delta ** 2, rate=1 / std_delta ** 2),
            reinterpreted_batch_ndims=2)  , ## gamma_delta
        lambda gamma_delta, std_delta, e_eta, nu: tfd.Independent(
            tfd.MultivariateNormalDiag(
                loc=nu + tf.matmul(e_eta * alpha_X * tf.matmul(tf.cast(source, tf.float32),tf.transpose(gamma_delta)), 
                                   tf.matmul(self.beta, self.bcumsum)) ,
                scale_diag=scale_diag),
            reinterpreted_batch_ndims=1)   #Y=
         ])
        model._to_track = self
        return model

### generate sample feature data
num_targets=4
num_subjects=10
num_sources=3

X=tf.random.uniform(shape=[num_subjects,4])
source=tf.one_hot(tf.random.categorical(tf.math.log([[0.5, 0.3, 0.2]]), num_subjects),3)[0,:]
treatment_group= tf.transpose(tf.random.categorical(tf.math.log([[0.5, 0.5]]), num_subjects))

X.shape, source.shape, treatment_group.shape

### get model from the features
WLSclass=BayesianWLS(num_targets,num_subjects,num_sources)
WLSmodel=WLSclass(X,source,treatment_group)

print(WLSmodel.resolve_graph())
WLSmodel.event_shape_tensor()

### It's ok to get one sample from the model
WLSmodel.sample()

### It's NOT allowed to get two samples from the model. How is the sample number working in the model?
WLSmodel.sample(2)

**Comment by davmre** [comment]
You don't say what error you're getting, but generally there are quite a number of ways to misspecify shapes in JointDistribution models. Most of these would be cleared up by using `JointDistributionSequentialAutoBatched` in place of `JointDistributionSequential` --- have you tried doing that?

**Comment by xiaolong1979** [comment]
Hi, davmre, 
Thanks for your help. It appears JointDistributionSequentialAutoBatched works better. There are a lot of warnings, like WARNING:tensorflow:Using a while_loop for converting StatelessRandomGetAlg.

Here are the codes I used. Is there any place that should be fixed?

### generate sample feature data
num_targets=4
num_subjects=10
num_sources=3

X=tf.random.uniform(shape=[num_subjects,4])
source=tf.one_hot(tf.random.categorical(tf.math.log([[0.5, 0.3, 0.2]]), num_subjects),3)[0,:]
treatment_group= tf.transpose(tf.random.categorical(tf.math.log([[0.5, 0.5]]), num_subjects))

### get feature and treatment fixed effect 
Bcov=tf.concat([tf.cast(treatment_group, tf.float32), X],axis=1)
alpha_X=tf.math.exp(tf.keras.layers.Dense(1)(Bcov)) ### exp(alpha*X)

beta =tf.Variable(tf.ones(shape=(1,num_targets),dtype=tf.float32), shape=(1,num_targets), name="beta") #  beta in eq
c_ones=int(num_targets*(num_targets+1)/2) ### for triangle matrix with 1
bcumsum=tf.dtypes.cast(tf.transpose(tfp.math.fill_triangular([1]*c_ones)), tf.float32)

## get treatment and source specific error stds
tx_source=tf.concat([tf.cast(treatment_group, tf.float32), tf.cast(source, tf.float32)],axis=1)
scale_diag=tf.repeat(tf.exp(tf.matmul(tx_source,tf.ones([1+num_sources, 1]))), num_targets, axis=1)

def model():
  nu = yield tfd.MultivariateNormalDiag(loc=tf.zeros(num_subjects), scale_identity_multiplier=1.0)
  eta = yield tfd.Gamma(concentration=1/(1.2)**2+tf.zeros(num_subjects), rate=1/(1.2)**2+tf.zeros(num_subjects))
  std_delta = yield tfd.Uniform(low=0.1+tf.zeros(num_sources), high=1.3+tf.zeros(num_sources))
  delta = yield tfd.Gamma(concentration=1 / std_delta ** 2, rate=1 / std_delta ** 2)
  Y = yield tfd.Independent(
            tfd.MultivariateNormalDiag(
            loc=tf.repeat(tf.expand_dims(nu, -1), repeats=num_targets, axis=-1)
           + tf.repeat(tf.expand_dims(tf.exp(eta), -1), repeats=num_targets, axis=-1)
           *tf.repeat(tf.expand_dims(tf.exp(tf.tensordot( delta,source, axes=((0),(1)))), -1), repeats=num_targets, axis=-1)
           *tf.repeat(tf.expand_dims(tf.transpose(alpha_X), -1), repeats=num_targets, axis=-1)*tf.tensordot(beta,bcumsum, axes=((1), (0)))
           , 
            scale_diag=scale_diag)  ,
              reinterpreted_batch_ndims=1)
joint = tfd.JointDistributionCoroutineAutoBatched(model)

print(joint.event_shape_tensor())
print(joint.batch_shape)
# joint.sample(3)
joint.log_prob(joint.sample(3))
